### PR TITLE
[INCOMPLETE] Tutorial Map Button

### PR DESCRIPTION
**Pull request by RoiEXLab**
This PR builds on top of #693 because NewGameChooserMapCache is getting removed there.
Just treat this PR as an open Issue...
This is supposed to adress #438

This PR is almost finished, but for whatever reason `getResourcePUs(gameData)` is returning null / `tutorialEntry.getGameData().getResourceList().getResource(Constants.PUS)` is `null` the gameData is retrieved via `new NewGameChooserModel().findByName("Tutorial").getGameData();`

As already posted in #438 here's the full StackTrace:

```
java.lang.NullPointerException
	at games.strategy.triplea.ui.AbstractStatPanel$ResourceStat.getName(AbstractStatPanel.java:71)
	at games.strategy.triplea.ui.StatPanel$StatTableModel.getColumnName(StatPanel.java:258)
	at javax.swing.JTable.addColumn(Unknown Source)
	at javax.swing.JTable.createDefaultColumnsFromModel(Unknown Source)
	at javax.swing.JTable.tableChanged(Unknown Source)
	at javax.swing.JTable.setModel(Unknown Source)
	at javax.swing.JTable.<init>(Unknown Source)
	at javax.swing.JTable.<init>(Unknown Source)
	at games.strategy.triplea.ui.StatPanel$1.<init>(StatPanel.java:72)
	at games.strategy.triplea.ui.StatPanel.initLayout(StatPanel.java:72)
	at games.strategy.triplea.ui.StatPanel.<init>(StatPanel.java:64)
	at games.strategy.triplea.ui.TripleAFrame.<init>(TripleAFrame.java:340)
	at games.strategy.triplea.TripleA.lambda$0(TripleA.java:130)
	at games.strategy.common.swing.SwingAction.lambda$0(SwingAction.java:61)
	at java.awt.event.InvocationEvent.dispatch(Unknown Source)
	at java.awt.EventQueue.dispatchEventImpl(Unknown Source)
	at java.awt.EventQueue.access$500(Unknown Source)
	at java.awt.EventQueue$3.run(Unknown Source)
	at java.awt.EventQueue$3.run(Unknown Source)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(Unknown Source)
	at java.awt.EventQueue.dispatchEvent(Unknown Source)
	at java.awt.EventDispatchThread.pumpOneEventForFilters(Unknown Source)
	at java.awt.EventDispatchThread.pumpEventsForFilter(Unknown Source)
	at java.awt.EventDispatchThread.pumpEventsForHierarchy(Unknown Source)
	at java.awt.EventDispatchThread.pumpEvents(Unknown Source)
	at java.awt.EventDispatchThread.pumpEvents(Unknown Source)
	at java.awt.EventDispatchThread.run(Unknown Source)
```

I have absolutely no idea whats causing this problem...

BTW 90% of the changes I made are in MetaSetupPanel.java most of the other stuff is from #693

---
This change is [Reviewable](https://reviewable.io/reviews/triplea-game/triplea/726)

### CODE REVIEW

File: src/games/strategy/triplea/TripleA.java [hunk 1]
```diff
@@ -137,6 +138,9 @@ public void startGame(final IGame game, final Set<IGamePlayer> players, final bo
         connectPlayers(players, frame);
         frame.setExtendedState(Frame.MAXIMIZED_BOTH);
         frame.toFront();
+        }catch(Exception e){
+          e.printStackTrace();
+        }
       });
```
Comment (RoiEXLab) [inline src/games/strategy/triplea/TripleA.java]: Just for debugging... Will probably get removed when this is working...

**Comment by RoiEXLab** [comment]
Side Note here:
If this is working some time making it shiny the first tim ea user opens TripleA will be no big deal...

File: src/games/strategy/engine/framework/startup/ui/MetaSetupPanel.java [hunk 2]
```diff
@@ -124,55 +137,35 @@ private void layoutComponents() {
...
+    for(PlayerID player : tutorialEntry.getGameData().getPlayerList()){
+      playerTypes.put(player.getName(),
+          player.getName().startsWith("AI") || player.getName().startsWith("Neutral") ? "Hard (AI)" : "Human");
+    }
+    final PlayerListing playerListing =
+        new PlayerListing(null, playersEnabled, playerTypes, gameSelectorModel.getGameData().getGameVersion(),
+            gameSelectorModel.getGameName(), gameSelectorModel.getGameRound(), null, null);
+    ILauncher tutorialLauncher = new LocalLauncher(gameSelectorModel, new PlainRandomSource(), playerListing);
+    tutorialLauncher.launch(parent);
+    System.out.println("PUS" + (tutorialEntry.getGameData().getResourceList().getResource(Constants.PUS) != null));
```
Comment (DanVanAtta) [inline src/games/strategy/engine/framework/startup/ui/MetaSetupPanel.java]: The tutorial game does launch. I would check the code path that launches the game normally, to see if there is something different there. My thought was that may PU was missing from the XML spec, but given the normal launch, probably not the case. You could add in some temporary "if (!null)" type of checksk to get past the issue temporarily.

**Comment by RoiEXLab** [comment]
@DanVanAtta 
In a class - I don't remeber what it's called - there is a HashMap with strings(pus etc.) being the keys and returning resource object as values... well apparently launching like I do makes this map being left empty so null gets returned... I tried replacing the HashMap with my own "NullHashMap" (a subclass of HashMap) throwing Exceptions when get() returns null - well it turns out, that when starting tripleA and the maps get loaded initially exceptions were thrown all over the place... so no real solution.
I did a null check and defaulted to an empty string...
Guess what?
```
java.lang.IllegalArgumentException: Attempt to access invalid state: 0
	at games.strategy.engine.data.GameSequence.getStep(GameSequence.java:144)
	at games.strategy.engine.data.GameSequence.getStep(GameSequence.java:138)
	at games.strategy.triplea.ui.TripleAFrame.updateStep(TripleAFrame.java:1429)
	at games.strategy.triplea.ui.TripleAFrame.<init>(TripleAFrame.java:414)
	at games.strategy.triplea.TripleA.lambda$0(TripleA.java:130)
	at games.strategy.common.swing.SwingAction.lambda$0(SwingAction.java:61)
	at java.awt.event.InvocationEvent.dispatch(Unknown Source)
	at java.awt.EventQueue.dispatchEventImpl(Unknown Source)
	at java.awt.EventQueue.access$500(Unknown Source)
	at java.awt.EventQueue$3.run(Unknown Source)
	at java.awt.EventQueue$3.run(Unknown Source)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(Unknown Source)
	at java.awt.EventQueue.dispatchEvent(Unknown Source)
	at java.awt.EventDispatchThread.pumpOneEventForFilters(Unknown Source)
	at java.awt.EventDispatchThread.pumpEventsForFilter(Unknown Source)
	at java.awt.EventDispatchThread.pumpEventsForHierarchy(Unknown Source)
	at java.awt.EventDispatchThread.pumpEvents(Unknown Source)
	at java.awt.EventDispatchThread.pumpEvents(Unknown Source)
	at java.awt.EventDispatchThread.run(Unknown Source)
```

I have absolutely _nooo_ idea what my launcher setup method is doing other than the getLauncher method in the LocalSetupPanel class.
The weird this about this is, that technically this is a problem with the game data... but I'm getting the game data by the `new NewGameChooserModel().findByName("Tutorial").getGameData();` method - so I have no influence on that...
So there has to be a method call on a object missing...
I have no idea on where to start...

**Comment by ron-murhammer** [comment]
@RoiEXLab Looks like this needs rebased.

**Comment by RoiEXLab** [comment]
Closing - View #438 for details

**Pull request closed without merging**